### PR TITLE
Aligns births deaths pregnancy

### PR DESCRIPTION
**Pull request by kaa2102**
### Checklist
- [x] Code commented & docstrings added
- [ ] New tests were needed and have been added
- [ ] A new version number was needed & changelog has been updated
- [ ] A new PyPI version needs to be released

### CODE REVIEW

**Comment by kaa2102** [comment]
The PR changes the way Births are generated (from deterministic to stochastic and tied to Mothers to be CRN-safe). 2 tests fail in test_demographics.py --> test_nigeria() and test_constant_pop() due to the changes in the birth rate. Changing the death_rates and tolerances in the np.isclose() functions from 5% to 10-20%.

**Review by RomeshA — COMMENTED** [review]
I'm probably missing some of the context for the changes here and the specific issue being addressed - but curious how this will interact with the maternal network, which also tracks the relationships between mothers and unborn children? Is the plan to keep both of them, or to move tracking it directly in the `People` object?

File: starsim/people.py [hunk 1]
```diff
@@ -61,6 +61,9 @@ def __init__(self, n_agents, age_data=None, extra_states=None):
         states = [
             ss.BoolArr('alive', default=True),  # Time index for death
             ss.BoolArr('female', default=ss.bernoulli(name='female', p=0.5)),
+            ss.Arr('parent_uids', dtype=[]), # List indicating the UID of mother giving birth
```
[the remark this answers is not in the document]
Comment (kaa2102) [reply starsim/people.py:64]: I had a chat with @cliffckerr on the approach and I am going to refactor the parent_uids `Arr` as an `IndexArr`  instead.

File: starsim/demographics.py [hunk 2]
```diff
@@ -83,27 +90,47 @@ def update(self):
...
+        # People eligible to become pregnant. We don't remove pregnant people here, these
+        # are instead handled in the fertility_dist logic as the rates need to be adjusted
+        denom_conds = (sim.people.female) & (sim.people.alive) & (sim.people.age>=18)
+        inds_to_choose_from = sim.people.uid[denom_conds]
+        
+        self.pars.birth_rate_data = ss.bernoulli(p=scaled_birth_prob)
+        self.pars.birth_rate_data.initialize(trace='any', sim=sim, slots=None)
+        mother_uids = self.pars.birth_rate_data.filter(inds_to_choose_from)
+        n_new=len(mother_uids)
+        np.array(mother_uids)
```
[the remark this answers is not in the document]
Comment (kaa2102) [reply starsim/demographics.py:119]: The `n_new` line is a piece of legacy code used to return the number of new uids (babies) to be added using `sim.people.grow(n_new)`.

File: starsim/demographics.py [hunk 3]
```diff
@@ -62,8 +62,15 @@ def initialize(self, sim):
             br_val = self.pars.birth_rate[self.metadata.data_cols['cbr']]
             all_birth_rates = np.interp(sim.yearvec, br_year, br_val)
             self.pars.birth_rate = all_birth_rates
-        return
 
+        '''Empty the parent and child uids on first sim iteration sim.ti==0 '''
+        if(sim.ti==0):
+            sim.people.parent_uids=[]
```
[the remark this answers is not in the document]
Comment (kaa2102) [reply starsim/demographics.py:68]: I am going to remove this code in the refactor from `Arr` to `IndexArr`.

**Review by daniel-klein — CHANGES_REQUESTED** [review]
Agree with the comments from @RomeshA and @cliffckerr that this PR needs to be refactored. As we discussed previously, we should not attempt to track every child a partner has, as that would be a one to many relationship, and instead only track a single parent for each newborn - a one to one relationship. This change should simplify the implementation and allow conformation with typical usage patterns of state array.

**Pull request closed without merging**